pkg/ratelimit/option.go

```diff
@@ -16,39 +16,112 @@ package ratelimit
 
 import "golang.org/x/time/rate"
 
+// UpdateStatus is flags for updating limiter config.
+type UpdateStatus uint32
+
+// Flags for limiter.
+const (
+	// QPSNoChange shows that limiter's config isn't changed.
+	QPSNoChange UpdateStatus = 1 << iota
+	// QPSChanged shows that limiter's config is changed and not deleted.
+	QPSChanged
+	// QPSDeleted shows that limiter's config is deleted.
+	QPSDeleted
+	// ConcurrencyNoChange shows that limiter's config isn't changed.
+	ConcurrencyNoChange
+	// ConcurrencyChanged shows that limiter's config is changed and not deleted.
+	ConcurrencyChanged
+	// ConcurrencyDeleted shows that limiter's config is deleted.
+	ConcurrencyDeleted
+	// InAllowList shows that limiter's config isn't changed because it is in in allow list.
+	InAllowList
+	// Ignore shows that the status can be ignored when initiated
+	Ignore
+)
+
 // Option is used to create a limiter with the optional settings.
 // these setting is used to add a kind of limiter for a service
-type Option func(string, *Limiter)
+type Option func(string, *Limiter) UpdateStatus
 
 // AddLabelAllowList adds a label into allow list.
 // It means the given label will not be limited
 func AddLabelAllowList() Option {
-	return func(label string, l *Limiter) {
+	return func(label string, l *Limiter) UpdateStatus {
 		l.labelAllowList[label] = struct{}{}
+		return Ignore
+	}
+}
+
+func updateConcurrencyConfig(l *Limiter, label string, limit uint64) UpdateStatus {
+	l.configMux.Lock()
+	defer l.configMux.Unlock()
+
+	cfg := l.labelConfig[label]
+	if cfg.ConcurrencyLimit == limit {
```

Comment:
Is it possible that the `label` does not exist, so the `cfg.ConcurrencyLimit` is always equal to a zero limit which will make the deleting can not be performed forever.

Reply:
> Is it possible that the `label` does not exist, so the `cfg.ConcurrencyLimit` is always equal to a zero `limit` which will make the deleting can not be performed forever.

I think if the `label` does not exist and the `limit` is equal to 0, then the `label` has no limiter, so it does not need to be deleted and there is no change.

Reply:
Init the labelConfig directly when it not exist will make me(the reviewer) happier  I think😄